convert/convert.go

```diff
@@ -0,0 +1,144 @@
+package convert
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+)
+
+const SegmentSize = 32 * 1024 * 1024
+const BufferSize = 1024 * 1024
+const Workers = 8
+
+type Options struct {
+	SegmentSize int64
+	BufferSize  int
+	Workers     int
+}
+
+type Converter struct {
+	// Read only after starting.
+	size        int64
+	segmentSize int64
+	bufferSize  int
+	workers     int
+
+	// State modified during Convert, protected by the mutex.
+	mutex  sync.Mutex
+	offset int64
+	err    error
+}
+
+// New return a new converter intialized from options.
+func New(opts Options) *Converter {
+	if opts.SegmentSize == 0 {
+		opts.SegmentSize = SegmentSize
+	}
+	if opts.BufferSize == 0 {
+		opts.BufferSize = BufferSize
+	}
+	if opts.Workers == 0 {
+		opts.Workers = Workers
+	}
+	return &Converter{
+		segmentSize: opts.SegmentSize,
+		bufferSize:  opts.BufferSize,
+		workers:     opts.Workers,
+	}
+}
```

Comment:
This should validate opts (e.g. negative integers) and return err

Reply:
Right, will add validation.

Reply:
Addressed in current version.

Validation extracted to Options.Validate() that can also be used by the caller to validated options before creating a converter.